File: ts/core/guardrails/tradeManagement.ts
export type Side = "long" | "short";

export interface TradeLike {
  id: string;
  symbol: string;
  side: Side;
  entry: number;
  initialStop: number;
  currentStop?: number | null;
  size: number; // current open size (units/contracts)
}

export interface TradeManagementState {
  beApplied: boolean;
  bePlusApplied: boolean;
  tp1Done: boolean;
  runnerActive: boolean;
}

export const DEFAULT_TM_STATE: TradeManagementState = {
  beApplied: false,
  bePlusApplied: false,
  tp1Done: false,
  runnerActive: false
};

export interface TradeManagementParams {
  feeBufferR: number; // default 0.05R
  beTriggerR: number; // default 1.0R
  bePlusTriggerR: number; // default 1.5R
  bePlusLockR: number; // default 0.25R
  tp1R: number; // default 1.0R
  tp1ClosePct: number; // default 0.5 (50%)
  runnerTrailBufferR: number; // default 0.5R behind swing
}

export const DEFAULT_TM_PARAMS: TradeManagementParams = {
  feeBufferR: 0.05,
  beTriggerR: 1.0,
  bePlusTriggerR: 1.5,
  bePlusLockR: 0.25,
  tp1R: 1.0,
  tp1ClosePct: 0.5,
  runnerTrailBufferR: 0.5
};

export interface SwingPoint {
  // For long: last confirmed swingLow
  // For short: last confirmed swingHigh
  price: number;
  ts?: number;
}

export type MgmtActionReason =
  | "tp1_partial"
  | "be"
  | "be_plus"
  | "runner_trail";

export interface MgmtAction {
  reason: MgmtActionReason;
  // If defined, executor may move stop to this value (after validation with venue rules).
  newStop?: number;
  // If defined, executor may close this fraction of current position size (0..1).
  closePct?: number;
  // Helpful for logs
  profitR: number;
  oldStop?: number | null;
}

export interface MgmtResult {
  nextState: TradeManagementState;
  actions: MgmtAction[];
  profitR: number;
  R: number;
}

/**
 * 1R = abs(entry - initialStop)
 * profitR is signed so it is positive in profit direction for both long and short.
 */
export function computeR(trade: TradeLike): number {
  return Math.abs(trade.entry - trade.initialStop);
}

export function computeProfitR(trade: TradeLike, currentPrice: number): number {
  const R = computeR(trade);
  if (R === 0) return 0;

  if (trade.side === "long") return (currentPrice - trade.entry) / R;
  return (trade.entry - currentPrice) / R;
}

function tightenStopOnly(
  side: Side,
  currentStop: number | null | undefined,
  proposedStop: number
): number | undefined {
  if (currentStop == null) return proposedStop;

  // long: tighter stop means higher stop
  if (side === "long") return proposedStop > currentStop ? proposedStop : undefined;

  // short: tighter stop means lower stop
  return proposedStop < currentStop ? proposedStop : undefined;
}

function clampPct01(x: number): number {
  if (x < 0) return 0;
  if (x > 1) return 1;
  return x;
}

/**
 * Returns management actions (partial close + stop moves) and updated state.
 * This function does NOT place orders; it only returns instructions.
 */
export function evaluateTradeManagement(
  trade: TradeLike,
  tm: TradeManagementState,
  currentPrice: number,
  params: TradeManagementParams = DEFAULT_TM_PARAMS,
  swing?: SwingPoint
): MgmtResult {
  const R = computeR(trade);
  const profitR = computeProfitR(trade, currentPrice);

  const actions: MgmtAction[] = [];
  const oldStop = trade.currentStop ?? null;

  // If R is invalid, do nothing safely.
  if (R <= 0 || !Number.isFinite(R)) {
    return { nextState: tm, actions, profitR, R };
  }

  // --- Block B: TP1 partial + runner creation
  let nextState: TradeManagementState = { ...tm };

  if (!nextState.tp1Done && profitR >= params.tp1R) {
    actions.push({
      reason: "tp1_partial",
      closePct: clampPct01(params.tp1ClosePct),
      profitR,
      oldStop
    });

    nextState.tp1Done = true;
    nextState.runnerActive = true;
  }

  // --- Block C: Apply BE
  // Stop = entry + sign * (feeBufferR * R)
  if (!nextState.beApplied && profitR >= params.beTriggerR) {
    const sign = trade.side === "long" ? +1 : -1;
    const proposed = trade.entry + sign * (params.feeBufferR * R);

    const tightened = tightenStopOnly(trade.side, oldStop, proposed);
    if (tightened !== undefined) {
      actions.push({
        reason: "be",
        newStop: tightened,
        profitR,
        oldStop
      });
      nextState.beApplied = true;
    }
  }

  // --- Block D: Apply BE+
  // Stop = entry + sign * ((bePlusLockR + feeBufferR) * R)
  if (!nextState.bePlusApplied && profitR >= params.bePlusTriggerR) {
    const sign = trade.side === "long" ? +1 : -1;
    const proposed =
      trade.entry + sign * ((params.bePlusLockR + params.feeBufferR) * R);

    // use the most recent stop if we already planned a BE move
    const latestStop =
      actions
        .slice()
        .reverse()
        .find(a => a.newStop !== undefined)?.newStop ?? oldStop;

    const tightened = tightenStopOnly(trade.side, latestStop ?? null, proposed);
    if (tightened !== undefined) {
      actions.push({
        reason: "be_plus",
        newStop: tightened,
        profitR,
        oldStop: latestStop ?? oldStop
      });
      nextState.bePlusApplied = true;
    }
  }

  // --- Block E: Runner trail (swing mode)
  // Only when runner is active and after BE+ trigger level
  if (nextState.runnerActive && profitR >= params.bePlusTriggerR && swing) {
    const latestStop =
      actions
        .slice()
        .reverse()
        .find(a => a.newStop !== undefined)?.newStop ?? oldStop;

    let proposed: number;
    if (trade.side === "long") {
      proposed = swing.price - params.runnerTrailBufferR * R;
    } else {
      proposed = swing.price + params.runnerTrailBufferR * R;
    }

    const tightened = tightenStopOnly(trade.side, latestStop ?? null, proposed);
    if (tightened !== undefined) {
      actions.push({
        reason: "runner_trail",
        newStop: tightened,
        profitR,
        oldStop: latestStop ?? oldStop
      });
    }
  }

  return { nextState, actions, profitR, R };
}

File: py/core/guardrails/trade_management.py
from __future__ import annotations
from dataclasses import dataclass
from typing import List, Optional, Dict, Any

from .trade_types import Trade, Action, profit_in_R, risk_R

@dataclass
class TMParams:
    tp1_r: float = 1.0
    tp1_close_pct: float = 0.5
    be_offset_r: float = 0.05      # 0.5R would be 0.5, but your example used 0.05? No: your example is 0.5R (entry 100, stop 90 => R=10, +0.5R=+5 => 105). You had 100.5 so offset was 0.05R. We'll keep it explicit and set it in the test.
    be_plus_r: float = 1.5
    be_plus_stop_r: float = 0.3    # same idea: set explicitly to match your TS test. Example shows +3 on entry 100 -> 103 => 0.3R when R=10.
    runner_trail_start_r: float = 2.0
    runner_trail_step_r: float = 0.5  # move stop each additional 0.5R
    runner_trail_lock_r: float = 1.0  # how much profit to lock once trailing

@dataclass
class TMState:
    did_tp1: bool = False
    did_be: bool = False
    did_be_plus: bool = False
    runner_mode: bool = False
    last_trail_level_r: float = 0.0

def _stop_from_entry(entry: float, r_value: float, side: str, stop_r: float) -> float:
    """
    stop = entry +/- stop_r*R depending on side.
    buy: stop = entry + stop_r*R (locks profit above entry)
    sell: stop = entry - stop_r*R
    """
    if side == "buy":
        return entry + (stop_r * r_value)
    return entry - (stop_r * r_value)

def evaluate_trade_management(
    trade: Trade,
    state: TMState,
    price: float,
    params: TMParams
) -> List[Action]:
    """
    Returns a list of actions that should happen at this price snapshot.
    """
    actions: List[Action] = []

    base_stop = trade.current_stop if trade.current_stop is not None else trade.initial_stop
    r_value = risk_R(trade.entry, trade.initial_stop, trade.side)
    if r_value <= 0:
        return actions

    pr = profit_in_R(trade.entry, trade.initial_stop, trade.side, price)

    # TP1 + BE at +1R
    if (not state.did_tp1) and pr >= params.tp1_r:
        actions.append(Action(
            reason="tp1_partial",
            close_pct=params.tp1_close_pct,
            profit_r=pr,
            old_stop=base_stop
        ))
        state.did_tp1 = True

    if (state.did_tp1 and (not state.did_be)) and pr >= params.tp1_r:
        new_stop = _stop_from_entry(trade.entry, r_value, trade.side, params.be_offset_r)
        old = trade.current_stop if trade.current_stop is not None else trade.initial_stop
        trade.current_stop = new_stop
        actions.append(Action(
            reason="be",
            new_stop=new_stop,
            old_stop=old,
            profit_r=pr
        ))
        state.did_be = True

    # BE+ at +1.5R
    if (not state.did_be_plus) and pr >= params.be_plus_r:
        new_stop = _stop_from_entry(trade.entry, r_value, trade.side, params.be_plus_stop_r)
        old = trade.current_stop if trade.current_stop is not None else trade.initial_stop
        trade.current_stop = new_stop
        actions.append(Action(
            reason="be_plus",
            new_stop=new_stop,
            old_stop=old,
            profit_r=pr
        ))
        state.did_be_plus = True
        state.runner_mode = True

    # Runner trail
    if state.runner_mode and pr >= params.runner_trail_start_r:
        # Determine the trail step "level"
        level = pr - params.runner_trail_start_r
        # how many steps of runner_trail_step_r have we advanced?
        steps = int(level // params.runner_trail_step_r)
        target_level_r = params.runner_trail_start_r + (steps * params.runner_trail_step_r)

        if target_level_r > state.last_trail_level_r:
            # lock runner_trail_lock_r behind current achieved level
            lock_stop_r = target_level_r - params.runner_trail_lock_r
            new_stop = _stop_from_entry(trade.entry, r_value, trade.side, lock_stop_r)
            old = trade.current_stop if trade.current_stop is not None else trade.initial_stop

            # Only tighten stop in the right direction
            if trade.side == "buy":
                should_update = (trade.current_stop is None) or (new_stop > trade.current_stop)
            else:
                should_update = (trade.current_stop is None) or (new_stop < trade.current_stop)

            if should_update:
                trade.current_stop = new_stop
                actions.append(Action(
                    reason="runner_trail",
                    new_stop=new_stop,
                    old_stop=old,
                    profit_r=pr
                ))
                state.last_trail_level_r = target_level_r

    return actions
